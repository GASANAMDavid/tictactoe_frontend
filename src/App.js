import React, { useState } from "react";
import CreateGame from "./components/CreateGame";
import { getTranslations } from "./actions/translate";
import { createGame } from "./actions/createGame";
import { applyMove } from "./actions/applyMove";
import Board from "./components/Board";

const App = () => {
  let errorMessage = "";
  const [gameCreated, setGameCreated] = useState({
    id: null,
    board: [["-"], ["-"]],
    state: ""
  });

  const [gameInfo, setGameInfo] = useState({
    playerName: "",
    symbol: "",
    boardSize: null,
    gameMode: null,
    language: null,
  });

  const [translations, setTranslations] = useState({
    welcomeMessage: "Welcome to TicTacToe Game",
  });

  const getSelectedLanguageContent = (lang) => {
    getTranslations(lang).then((data) => {
      setTranslations(data);
    });
  };

  const handleInvalidMove = () => {
    errorMessage = "Invalid move";
  };

  const handleCreateGame = () => {
    let id;
    let board;
    createGame(gameInfo).then((createdGame) => {
      setGameCreated((({ id, board } = createdGame), { id, board }));
    });
  };

  const handleCellClick = (cellPosition) => {
    applyMove(gameCreated.id, cellPosition).then((response) => {
      setGameCreated({ ...gameCreated, ...response });
    });
  };

  return (
    <div className='container'>
      <div className='header'>
        {gameCreated.id
          ? translations.playGameHeader
          : translations.welcomeMessage}
      </div>
      <div className='content'>
        {gameCreated.id ? (
          <Board
            board={gameCreated.board}
            onCellClick={handleCellClick}
            onInvalidMove={handleInvalidMove}
            isOngoing={gameCreated.ongoing}
          />
        ) : (
          <CreateGame
            onSelectLanguage={getSelectedLanguageContent}
            onCreateGame={handleCreateGame}
            translations={translations}
            gameInfo={gameInfo}
            setGameInfo={setGameInfo}
          />
        )}
      </div>
      <div className='game-info'>
        <h4> Game Information </h4>
        <div>
          name:
          {gameInfo.playerName}
        </div>
        <div>
          Symbol:
          {gameInfo.symbol}
        </div>
        <div>
          game mode:
          {gameInfo.gameMode}
        </div>
        <div>
          opponent:
          {gameCreated.opponent}
        </div>
        <div>
          opponent symbol:
          {gameCreated.opponentSymbol}
        </div>
        <div></div>
      </div>
      <div className='status'>
        { errorMessage ? errorMessage : gameCreated.state}
      </div>
    </div>
  );
};
export default App;
