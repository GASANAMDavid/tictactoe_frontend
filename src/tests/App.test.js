import React from "react";
import { shallow } from "enzyme";
import App from "../App";

describe("App", () => {
  let app;
  const mockDefaultGameState = {
    id: null,
    board: [["-"], ["-"]],
    state: {
      message: "",
      ongoing: true,
    },
  };

  describe("when game is not created yet", () => {
    beforeEach(() => {
      app = shallow(<App gameCreated={mockDefaultGameState} />);
    });
    it("renders the 'CreateGame' component", () => {
      expect(app.find("CreateGame").exists()).toBe(true);
    });
  });

  describe("when the game is created", () => {
    const newCreatedGameState = {
      id: 1,
      board: [["-"], ["-"]],
      state: {
        message: "",
        ongoing: true,
      },
    };

    beforeEach(() => {
      app = shallow(<App gameCreated={newCreatedGameState} />);
    });

    it("renders the 'Board' component", () => {
      expect(app.find("Board").exists()).toBe(true);
    });
  });
});
