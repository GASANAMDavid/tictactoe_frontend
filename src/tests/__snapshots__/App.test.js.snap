// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="container"
>
  <h3
    className="header"
  >
    Welcome to TicTacToe Game
  </h3>
  <CreateGame
    gameInfo={
      Object {
        "boardSize": 3,
        "gameMode": 1,
        "language": null,
        "playerName": "",
        "symbol": "",
      }
    }
    onCreateGame={[Function]}
    onSelectLanguage={[Function]}
    setGameInfo={[Function]}
    translations={
      Object {
        "welcomeMessage": "Welcome to TicTacToe Game",
      }
    }
  />
</div>
`;
